File: src/collision.js
/*
 * Copyright 2026 Manifold Tech Ltd.
 * Author: MENG Guotao <mengguotao@manifoldtech.cn>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Octree-based collision detection for point cloud data.
 * Builds a spatial index from Gaussian center positions and provides
 * sphere-vs-pointcloud queries for drone collision.
 */

const MAX_POINTS_PER_NODE = 64;
const MAX_DEPTH = 12;

class OctreeNode {
    constructor(minX, minY, minZ, maxX, maxY, maxZ, depth) {
        this.minX = minX; this.minY = minY; this.minZ = minZ;
        this.maxX = maxX; this.maxY = maxY; this.maxZ = maxZ;
        this.depth = depth;
        this.children = null; // null = leaf
        this.pointIndices = [];
    }

    get midX() { return (this.minX + this.maxX) * 0.5; }
    get midY() { return (this.minY + this.maxY) * 0.5; }
    get midZ() { return (this.minZ + this.maxZ) * 0.5; }

    intersectsSphere(cx, cy, cz, r) {
        // Closest point on AABB to sphere center
        const dx = Math.max(this.minX - cx, 0, cx - this.maxX);
        const dy = Math.max(this.minY - cy, 0, cy - this.maxY);
        const dz = Math.max(this.minZ - cz, 0, cz - this.maxZ);
        return (dx * dx + dy * dy + dz * dz) <= (r * r);
    }
}

export class Octree {
    constructor() {
        this.root = null;
        this.positions = null;
        this.pointCount = 0;
    }

    /**
     * Build the octree from a Float32Array of positions [x0,y0,z0, x1,y1,z1, ...]
     */
    build(positions, bounds) {
        this.positions = positions;
        this.pointCount = positions.length / 3;

        // Add small padding to bounds
        const pad = 0.1;
        this.root = new OctreeNode(
            bounds.min[0] - pad, bounds.min[1] - pad, bounds.min[2] - pad,
            bounds.max[0] + pad, bounds.max[1] + pad, bounds.max[2] + pad,
            0
        );

        // Insert all points
        const indices = [];
        for (let i = 0; i < this.pointCount; i++) {
            indices.push(i);
        }
        this._subdivide(this.root, indices);

        return this;
    }

    _subdivide(node, indices) {
        if (indices.length <= MAX_POINTS_PER_NODE || node.depth >= MAX_DEPTH) {
            node.pointIndices = indices;
            return;
        }

        node.children = new Array(8);
        const mx = node.midX, my = node.midY, mz = node.midZ;
        const childBounds = [
            [node.minX, node.minY, node.minZ, mx, my, mz],
            [mx, node.minY, node.minZ, node.maxX, my, mz],
            [node.minX, my, node.minZ, mx, node.maxY, mz],
            [mx, my, node.minZ, node.maxX, node.maxY, mz],
            [node.minX, node.minY, mz, mx, my, node.maxZ],
            [mx, node.minY, mz, node.maxX, my, node.maxZ],
            [node.minX, my, mz, mx, node.maxY, node.maxZ],
            [mx, my, mz, node.maxX, node.maxY, node.maxZ],
        ];

        const childIndices = [[], [], [], [], [], [], [], []];

        for (const idx of indices) {
            const x = this.positions[idx * 3];
            const y = this.positions[idx * 3 + 1];
            const z = this.positions[idx * 3 + 2];
            const ci = (x >= mx ? 1 : 0) | (y >= my ? 2 : 0) | (z >= mz ? 4 : 0);
            childIndices[ci].push(idx);
        }

        for (let i = 0; i < 8; i++) {
            const b = childBounds[i];
            node.children[i] = new OctreeNode(b[0], b[1], b[2], b[3], b[4], b[5], node.depth + 1);
            if (childIndices[i].length > 0) {
                this._subdivide(node.children[i], childIndices[i]);
            }
        }
    }

    /**
     * Collect AABBs of all non-empty leaf nodes.
     * Returns a Float32Array: [minX,minY,minZ, maxX,maxY,maxZ, ...] (6 floats per leaf).
     * Result is cached after first call.
     */
    getLeafBounds() {
        if (this._leafBoundsCache) return this._leafBoundsCache;
        const list = [];
        if (this.root) this._collectLeaves(this.root, list);
        this._leafBoundsCache = new Float32Array(list);
        return this._leafBoundsCache;
    }

    _collectLeaves(node, list) {
        if (node.children) {
            for (const child of node.children) {
                if (child) this._collectLeaves(child, list);
            }
        } else if (node.pointIndices.length > 0) {
            list.push(node.minX, node.minY, node.minZ, node.maxX, node.maxY, node.maxZ);
        }
    }

    /**
     * Query all points within a sphere. Returns array of { index, x, y, z, distSq }.
     */
    querySphere(cx, cy, cz, radius) {
        const results = [];
        const rSq = radius * radius;
        if (this.root) {
            this._querySphereNode(this.root, cx, cy, cz, radius, rSq, results);
        }
        return results;
    }

    _querySphereNode(node, cx, cy, cz, r, rSq, results) {
        if (!node.intersectsSphere(cx, cy, cz, r)) return;

        if (node.children) {
            for (const child of node.children) {
                if (child) this._querySphereNode(child, cx, cy, cz, r, rSq, results);
            }
        } else {
            for (const idx of node.pointIndices) {
                const px = this.positions[idx * 3];
                const py = this.positions[idx * 3 + 1];
                const pz = this.positions[idx * 3 + 2];
                const dx = px - cx, dy = py - cy, dz = pz - cz;
                const distSq = dx * dx + dy * dy + dz * dz;
                if (distSq <= rSq) {
                    results.push({ index: idx, x: px, y: py, z: pz, distSq });
                }
            }
        }
    }
}

/**
 * Collision detection result processor.
 * Given a sphere query result, computes collision normal and penetration.
 */
export function computeCollisionResponse(dronePos, radius, queryResults) {
    if (queryResults.length === 0) {
        return null;
    }

    // Compute average direction from nearby points to drone (surface normal estimate)
    let nx = 0, ny = 0, nz = 0;
    let minDist = Infinity;

    for (const pt of queryResults) {
        const dx = dronePos.x - pt.x;
        const dy = dronePos.y - pt.y;
        const dz = dronePos.z - pt.z;
        const dist = Math.sqrt(pt.distSq);
        if (dist < 0.0001) continue;

        // Weight by inverse distance (closer points contribute more to normal)
        const w = 1.0 / (dist + 0.001);
        nx += dx * w;
        ny += dy * w;
        nz += dz * w;

        if (dist < minDist) minDist = dist;
    }

    // Normalize
    const len = Math.sqrt(nx * nx + ny * ny + nz * nz);
    if (len < 0.0001) {
        // Degenerate case: push straight up
        return { normal: { x: 0, y: 1, z: 0 }, penetration: radius - minDist, pointCount: queryResults.length };
    }

    nx /= len; ny /= len; nz /= len;
    const penetration = radius - minDist;

    return {
        normal: { x: nx, y: ny, z: nz },
        penetration: Math.max(0, penetration),
        pointCount: queryResults.length
    };
}

File: native/src/collision.rs
/// Octree-based collision detection for point cloud data.
/// Ported from src/collision.js.

use cgmath::Vector3;

const MAX_POINTS_PER_NODE: usize = 64;
const MAX_DEPTH: u32 = 12;

struct OctreeNode {
    min: [f32; 3],
    max: [f32; 3],
    depth: u32,
    children: Option<Box<[OctreeNode; 8]>>,
    point_indices: Vec<u32>,
}

impl OctreeNode {
    fn new(min: [f32; 3], max: [f32; 3], depth: u32) -> Self {
        Self { min, max, depth, children: None, point_indices: Vec::new() }
    }

    fn mid(&self) -> [f32; 3] {
        [
            (self.min[0] + self.max[0]) * 0.5,
            (self.min[1] + self.max[1]) * 0.5,
            (self.min[2] + self.max[2]) * 0.5,
        ]
    }

    fn intersects_sphere(&self, cx: f32, cy: f32, cz: f32, r: f32) -> bool {
        let dx = (self.min[0] - cx).max(0.0f32).max(cx - self.max[0]);
        let dy = (self.min[1] - cy).max(0.0f32).max(cy - self.max[1]);
        let dz = (self.min[2] - cz).max(0.0f32).max(cz - self.max[2]);
        dx * dx + dy * dy + dz * dz <= r * r
    }
}

pub struct SphereHit {
    pub index: u32,
    pub x: f32,
    pub y: f32,
    pub z: f32,
    pub dist_sq: f32,
}

pub struct CollisionResult {
    pub normal: Vector3<f32>,
    pub penetration: f32,
}

pub struct Octree {
    root: Option<OctreeNode>,
    positions: Vec<f32>, // [x0,y0,z0, x1,y1,z1, ...]
    point_count: usize,
}

impl Octree {
    pub fn new() -> Self {
        Self { root: None, positions: Vec::new(), point_count: 0 }
    }

    /// Build from a flat f32 slice of positions [x,y,z,...] and AABB bounds.
    pub fn build(&mut self, positions: Vec<f32>, bounds_min: [f32; 3], bounds_max: [f32; 3]) {
        self.positions = positions;
        self.point_count = self.positions.len() / 3;

        let pad = 0.1;
        let mut root = OctreeNode::new(
            [bounds_min[0] - pad, bounds_min[1] - pad, bounds_min[2] - pad],
            [bounds_max[0] + pad, bounds_max[1] + pad, bounds_max[2] + pad],
            0,
        );

        let indices: Vec<u32> = (0..self.point_count as u32).collect();
        Self::subdivide(&mut root, &indices, &self.positions);
        self.root = Some(root);
    }

    fn subdivide(node: &mut OctreeNode, indices: &[u32], positions: &[f32]) {
        if indices.len() <= MAX_POINTS_PER_NODE || node.depth >= MAX_DEPTH {
            node.point_indices = indices.to_vec();
            return;
        }

        let [mx, my, mz] = node.mid();
        let mn = node.min;
        let mx_ = node.max;
        let child_bounds: [[f32; 6]; 8] = [
            [mn[0], mn[1], mn[2], mx, my, mz],
            [mx, mn[1], mn[2], mx_[0], my, mz],
            [mn[0], my, mn[2], mx, mx_[1], mz],
            [mx, my, mn[2], mx_[0], mx_[1], mz],
            [mn[0], mn[1], mz, mx, my, mx_[2]],
            [mx, mn[1], mz, mx_[0], my, mx_[2]],
            [mn[0], my, mz, mx, mx_[1], mx_[2]],
            [mx, my, mz, mx_[0], mx_[1], mx_[2]],
        ];

        let mut child_indices: [Vec<u32>; 8] = Default::default();

        for &idx in indices {
            let i = idx as usize;
            let x = positions[i * 3];
            let y = positions[i * 3 + 1];
            let z = positions[i * 3 + 2];
            let ci = (if x >= mx { 1 } else { 0 })
                | (if y >= my { 2 } else { 0 })
                | (if z >= mz { 4 } else { 0 });
            child_indices[ci].push(idx);
        }

        let mut children: [OctreeNode; 8] = std::array::from_fn(|i| {
            let b = &child_bounds[i];
            OctreeNode::new([b[0], b[1], b[2]], [b[3], b[4], b[5]], node.depth + 1)
        });

        for i in 0..8 {
            if !child_indices[i].is_empty() {
                Self::subdivide(&mut children[i], &child_indices[i], positions);
            }
        }

        node.children = Some(Box::new(children));
    }

    /// Query all points within a sphere.
    pub fn query_sphere(&self, cx: f32, cy: f32, cz: f32, radius: f32) -> Vec<SphereHit> {
        let mut results = Vec::new();
        let r_sq = radius * radius;
        if let Some(root) = &self.root {
            Self::query_sphere_node(root, cx, cy, cz, radius, r_sq, &self.positions, &mut results);
        }
        results
    }

    fn query_sphere_node(
        node: &OctreeNode, cx: f32, cy: f32, cz: f32,
        r: f32, r_sq: f32, positions: &[f32], results: &mut Vec<SphereHit>,
    ) {
        if !node.intersects_sphere(cx, cy, cz, r) { return; }

        if let Some(children) = &node.children {
            for child in children.iter() {
                Self::query_sphere_node(child, cx, cy, cz, r, r_sq, positions, results);
            }
        } else {
            for &idx in &node.point_indices {
                let i = idx as usize;
                let px = positions[i * 3];
                let py = positions[i * 3 + 1];
                let pz = positions[i * 3 + 2];
                let dx = px - cx;
                let dy = py - cy;
                let dz = pz - cz;
                let dist_sq = dx * dx + dy * dy + dz * dz;
                if dist_sq <= r_sq {
                    results.push(SphereHit { index: idx, x: px, y: py, z: pz, dist_sq });
                }
            }
        }
    }
}

/// Compute collision normal and penetration from sphere query results.
pub fn compute_collision_response(
    drone_x: f32, drone_y: f32, drone_z: f32,
    radius: f32,
    hits: &[SphereHit],
) -> Option<CollisionResult> {
    if hits.is_empty() { return None; }

    let mut nx = 0.0f32;
    let mut ny = 0.0f32;
    let mut nz = 0.0f32;
    let mut min_dist = f32::INFINITY;

    for pt in hits {
        let dist = pt.dist_sq.sqrt();
        if dist < 0.0001 { continue; }

        let dx = drone_x - pt.x;
        let dy = drone_y - pt.y;
        let dz = drone_z - pt.z;

        let w = 1.0 / (dist + 0.001);
        nx += dx * w;
        ny += dy * w;
        nz += dz * w;

        if dist < min_dist { min_dist = dist; }
    }

    let len = (nx * nx + ny * ny + nz * nz).sqrt();
    if len < 0.0001 {
        return Some(CollisionResult {
            normal: Vector3::new(0.0, 1.0, 0.0),
            penetration: (radius - min_dist).max(0.0),
        });
    }

    nx /= len; ny /= len; nz /= len;

    Some(CollisionResult {
        normal: Vector3::new(nx, ny, nz),
        penetration: (radius - min_dist).max(0.0),
    })
}
